pandas_gbq/gbq.py

```diff
@@ -837,68 +681,172 @@ def _get_credentials_file():
+        if dialect == 'legacy':
+            query_job.use_legacy_sql = True
+        elif dialect == 'standard':
+            query_job.use_legacy_sql = False
+
+        if configuration:
+            for setting, value in configuration.items():
+                setattr(query_job, setting, value)
+        return query_job
+
+    def sync_query():
+        query_job = client.run_sync_query(query,
+                                          query_parameters=query_parameters)
+        query_job = _set_common_query_settings(query_job)
+        if verbose:
+            print("Query running...")
+        query_job.run()
+        if not query_job._properties.get("jobComplete", False):
+            raise QueryTimeout("Sync query timed out")
+        if verbose:
+            print("Query done.")
+            if query_job._properties.get("cacheHit", False):
+                print("Cache hit.")
+            else:
+                bytes_billed = int(query_job._properties
+                                   .get("totalBytesProcessed", 0))
+                bytes_processed = int(query_job._properties
+                                      .get("totalBytesBilled", 0))
+                print("Total bytes billed (processed): %s (%s)" %
+                      (sizeof_fmt(bytes_billed), sizeof_fmt(bytes_processed)))
+            print("\nRetrieving results...")
+        return query_job, None
+
+    def async_query():
+        query_job = client.run_async_query(str(uuid.uuid4()),
+                                           query,
+                                           query_parameters=query_parameters)
+        query_job = _set_common_query_settings(query_job)
+        query_job.begin()
+        try:
+            query_results = query_job.results().fetch_data()
+        except AttributeError:
+            query_results = query_job.result().fetch_data()
+        if verbose:
+            print("Query running...")
+        _wait_for_job(query_job)
+        if verbose:
+            print("Query done.")
+            if query_job._properties["statistics"]["query"].get("cacheHit",
+                                                                False):
+                print("Cache hit.")
+            elif ("statistics" in query_job._properties and
+                    "query" in query_job._properties["statistics"]):
+                bytes_billed = int(query_job
+                                   ._properties["statistics"]["query"]
+                                   .get("totalBytesProcessed", 0))
+                bytes_processed = int(query_job
+                                      ._properties["statistics"]["query"]
+                                      .get("totalBytesBilled", 0))
+                print("Total bytes billed (processed): %s (%s)" %
+                      (sizeof_fmt(bytes_billed), sizeof_fmt(bytes_processed)))
+            print("\nRetrieving results...")
+        return query_results, query_job
+
+    def get_columns_schema(query_results):
```

Comment:
this is still 1 really really long function.

Reply:
I realize that it may look a little long, but it's no longer nor any more complex than the previous run_query function (it's actually 30 lines shorter). And if google-cloud handled timeouts in async queries correctly, this would be another 30 lines shorter and two functions less. I'm not sure why or how this function would be split up even more as it follows the same logic as the original run_query: you pass it a query and it handles all the client work and returns the data along with the schema.